haskell-api.md

```diff
@@ -117,6 +117,97 @@ possible. Issues include:
 If you are creating a new `PUT` consider if it could be expressed in a more
 granular `PATCH` semantic.
 
+### `PATCH` semantics
+
+Modifications to an existing resource should be made via a `PATCH` request.
+
+### Specific guidelines
+
+A well-formed `PATCH` endpoint
+
+- Must update an entity's `updatedAt` field, if present.
+- Should return the modified resource. The resource should be re-fetched for
+  ease of testing and to ensure the update was correctly persisted.
+- Must accept `null` to unset nullable fields, disallowing it for
+  non-nullables.
+- Must 404 if a resource with the route's identifier(s) does not exist.
+- Must not create resources (that's the role of `POST`).
+- Must allow for fields to be optional, even allowing for entirely empty `PATCH`
+  objects.
```

Comment:
Should this be a validation failure? We have a number of `PATCH` endpoints that validate whether the payload is not `{}` and throw if it is. That leads to a lot of noisy checking, but the supposed benefit is that it catches spurious API calls and logic errors. On the other hand validating `not {}` breaks robustness principles.

Reply:
random thought: If you allow empty, does `PATCH /x {}` become a synonym for `GET /x`.

Reply:
> Should this be a validation failure? We have a number of PATCH endpoints that validate whether the payload is not {} and throw if it is.

Gilli and I discussing that our present "no empty patch" validation is unnecessary is what prompted this Guide.

> On the other hand validating not {} breaks robustness principles

That is where I/we landed, hence this guide making the push that way.

Reply:
> random thought: If you allow empty, does PATCH /x {} become a synonym for GET /x

@eborden yep, I tried to note that in the PR description.